xbmc/utils/EGLUtils.cpp

```diff
@@ -596,3 +590,48 @@ bool CEGLContextUtils::TrySwapBuffers()
 
   return (eglSwapBuffers(m_eglDisplay, m_eglSurface) == EGL_TRUE);
 }
+
+int32_t CEGLContextUtils::GetBufferAge()
+{
+  int32_t bufferAge = 2;
+  if (CEGLUtils::HasExtension(m_eglDisplay, "EGL_KHR_partial_update"))
+    eglQuerySurface(m_eglDisplay, m_eglSurface, EGL_BUFFER_AGE_KHR, &bufferAge);
+  else if (CEGLUtils::HasExtension(m_eglDisplay, "EGL_EXT_buffer_age"))
+    eglQuerySurface(m_eglDisplay, m_eglSurface, EGL_BUFFER_AGE_EXT, &bufferAge);
+
+  return bufferAge;
+}
+
+bool CEGLContextUtils::HasDamagedRegionSupport()
+{
+  return CEGLUtils::HasExtension(m_eglDisplay, "EGL_KHR_partial_update");
+}
+
+void CEGLContextUtils::SetDamagedRegions(const CDirtyRegionList& dirtyRegions)
+{
+  if (!CEGLUtils::HasExtension(m_eglDisplay, "EGL_KHR_partial_update"))
+    return;
+
+  std::vector<EGLint> rects;
+  if (dirtyRegions.empty())
+  {
+    rects.push_back((EGLint)0);
+    rects.push_back((EGLint)0);
+    rects.push_back((EGLint)0);
+    rects.push_back((EGLint)0);
+  }
+  else
+  {
+    EGLint height = CServiceBroker().GetWinSystem()->GetHeight();
+    for (const auto& i : dirtyRegions)
+    {
+      if (i.IsEmpty())
+        continue;
+      rects.push_back((EGLint)i.x1);
+      rects.push_back(height - (EGLint)i.y2);
+      rects.push_back((EGLint)i.Width());
+      rects.push_back((EGLint)i.Height());
+    }
+  }
+  m_eglSetDamageRegionKHR(m_eglDisplay, m_eglSurface, rects.data(), rects.size() / 4);
```

Comment:
This could be a little bit faster but I can't test it lacking a GPU with `EGL_KHR_partial_update`.
```suggestion
  using Rect = std::array<EGLint, 4>;

  if (dirtyRegions.empty())
  {
    static Rect zeroRect{};
    m_eglSetDamageRegionKHR(m_eglDisplay, m_eglSurface, zeroRect.data(), zeroRect.size());
  }
  else
  {
    EGLint height = CServiceBroker().GetWinSystem()->GetHeight();
    std::vector<Rect> rects;
    rects.reserve(dirtyRegions.size());
    for (const auto& i : dirtyRegions)
    {
      if (i.IsEmpty())
        continue;
      rects.push_back({(EGLint)i.x1, height - (EGLint)i.y2, (EGLint)i.Width(), (EGLint)i.Height()});
    }
    m_eglSetDamageRegionKHR(m_eglDisplay, m_eglSurface, reinterpret_cast<EGLint*>(rects.data()),
                            rects.size());
  }
```

Reply:
I'll try it out when I have the time.